### ISSUE TITLE
Remove dependency on RxJS (aka observable library agnostic)

### ISSUE THREAD

**Issue by jayphelps** [redux-observable/redux-observable#132]
While we love RxJS, there are indeed a large number of competing stream libraries. The paradigm this library intends to provide isn't in fact tied to RxJS and so other "ports" of redux-observable will eventually continue to pop up. e.g. [redux-most](https://github.com/joshburgess/redux-most).

IMO we should at least consider, "what if redux-observable didn't actually depend on RxJS at all?".

We use a couple RxJS operators internally right now, but mostly out of convenience. At first blush it seems we could get away without it. We would want to take great care that we're not reinventing too much of RxJS--it may turn out that reducing the number of RxJS operators we use internally is enough to no longer make it a peer dependency and instead, depend on it explicitly as merely an implementation detail, though even this isn't ideal because all of RxJS v5 is still in our node_modules and at least one prominent Electron user has reimplemented redux-observable for v4 for that reason alone.

I think the biggest question is actually how the ergonomics to RxJS might suffer in exchange. I do want to lean on the proposed [ECMAScript TC39 Observables spec](https://github.com/tc39/proposal-observable) as the core, so regardless of the stream library you wanted to use, it would _have to support the Observable spec_, or there at least be a way for someone to convert between them. most.js, xstream, kefir already support it, with [bacon.js support in discussion](https://github.com/baconjs/bacon.js/issues/633).

There would also be a question of documentation complexities. I'm not sure how maintainable it is to make separate versions of the docs for each stream library. In practice, it would likely end up being "RxJS v5 examples are documented by us, PRs for others are happily welcome" until we could make relations with other stream library users who would be willing to join some sort of "redux-observable core team"

So I think in that case, everyone would want to use an adapter, even RxJS users.

``` js
import rxjsAdapter, { combineEpics } from 'redux-observable-rxjs';

const rootEpic = combineEpics(epic1, epic2);
const epicMiddleware = createEpicMiddleware(rootEpic, { adapter: rxjsAdapter });
```

This does indeed increase setup friction a bit. Is it worth it to unify the communities long term? Nearly everyone reading this is probably an RxJS user, so I implore you to have empathy for others who don't share you opinion of Rx and objectively consider whether a little extra friction will make redux-observable better for everyone.

---

Don't freak out or get your hopes up. This is just to gauge community thoughts at this point. 😄 We may in fact be jumping the gun on this, and perhaps someone has arguments to let the communities grow independently?

If you have no comments other than agreeing/disagreeing feel free to add your "reaction" directly on this post with 👍 👎  

Cc/ @joshburgess

**Comment by istarkov** [comment]
> though even this isn't ideal because all of RxJS v5 is still in our node_modules and at least one prominent Electron user has reimplemented redux-observable for v4 for that reason alone

Does electron not supports bundling? As for usual webapp additional dependency is not a big problem.

**Comment by jayphelps** [comment]
@istarkov I'm told Electron itself doesn't provide any JavaScript bundling and that's it up to devs to choose to use typical ones like webpack/browserify/etc and that many of them do not. I do indeed think it would be ideal if devs concerned with their Electron package size would use a JS bundler, but we can't force them to do so. I won't let it be the deciding factor for us, just something to keep in mind.

**Comment by staltz** [comment]
So, this sounds a lot like what we did in Cycle.js with the "Cycle Diversity" release. First we had Cycle.js with RxJS v4 and Motorcycle with most.js, while RxJS v5 was taking shape to be production-ready. We wanted to combine all these efforts together. We knew that we would use some stream library still internally in the framework/libraries. We weren't sure though if it would stay as RxJS v4 or RxJS v5 or most.js or something else. That was one of the reasons why we created xstream: to be a tiny and fast stream library with little overhead for the libraries.

For that reason, consider (just consider), internally implementing redux-observable with xstream. Most of the important operators are in place.

The solution we had was:

- Libraries (drivers) implemented internally with xstream
- Adapter X to convert between stream library X and "some generic stream abstraction that looks like ES Observable"
- One top-level library for each stream library: `@cycle/xstream-run`, `@cycle/rxjs-run`, etc

This worked ok because people had very little code to change in order to migrate: they just had to change `@cycle/core` to `@cycle/mylibrary-run`, basically. One trick we used for detecting stream libraries: attaching the adapter to each driver function (a "plugin") like this https://github.com/cyclejs/cyclejs/blob/master/dom/src/makeDOMDriver.ts#L70

In my personal experience and opinion, if I would have built Cycle.js starting from scratch today, I would only support xstream, just because adding support for N more stream libraries takes more time. But I do realize a lot of people love most.js and RxJS, so it's important for them. In some situations, most.js is a life saver for its hyper performance.

About ES Observable. Currently, most.js, RxJS v5 and xstream convert from and to ES Observable. However, not really. ES Observable spec itself is slightly worse than utter crap (let's be honest), with its lack of subscription and with cancelToken in the subscribe argument. Currently, most.js and xstream use the wise version of ES Observable from some months ago:
- https://github.com/cujojs/most/blob/master/src/observable/fromObservable.js#L17
- https://github.com/staltz/xstream/blob/master/src/core.ts#L110

So, when talking about using ES Observable as a common generic type, ask yourself, *which ES Observable*? :trollface: 

Then, recently I've been contemplating building "Cycle Unified", a version of Cycle.js where we only support xstream, but conversion to/from other stream libraries is transparent and automatic. This would make it possible to drop `@cycle/xstream-run`, `@cycle/rxjs-run` and only have `@cycle/run`, with some intelligent detections in place inside `@cycle/run`, using the fact that each stream library supports converting to/from the uncrappy ES Observable. https://github.com/cyclejs/cyclejs/issues/425

**Comment by jayphelps** [comment]
@staltz valuable insight! re: ES Observable version, it's certainly a moving target. To me, it seemed like the only target all these stream libraries can somewhat agree on, even if there will definitely be variations at the various stages of the spec.

Very much appreciate your input.

**Comment by rgbkrk** [comment]
> Does electron not supports bundling?

No. It's pure node.js, running in Chromium.

> If Electron users did bundle, it wouldn't just reduce package size, it would reduce their start times as well because they wouldn't have to have so many filesystem IO calls for require but I get that bundlers suck compared to the awesome dev experience of just node-style require

It continues to be very tempting for [nteract](https://github.com/nteract/nteract) to rely on webpack, our main hold up is making the project friendly for contribution. We already have a great deal of process around our native modules, which we recently made even harder on developers to make everything better for users. Introducing a bundler would worry me if it scares off more new contributors (we keep getting more, and we love it).

I too find the cancelToken in the es observable spec weird, but haven't really given time to understanding what brought that about.

At least for folks like me that have already adopted redux-observable, adding some extra ceremony is easy. We add it in, and continue spending most of our time working on the core of app and working with our epics.

As for new contributors, yeah... Any amount of extra pieces they need to understand (or perceive that they need to) makes it a little harder to adopt a library, cognitively. I do like the approach @staltz mentioned for creating packages that expose a nice interface for each observable lib.

**Comment by sedubois** [comment]
A question that [came](https://github.com/mxstbr/react-boilerplate/issues/1139#issuecomment-255736687) to mind when considering redux-observable is, how big of a webpack bundle will it produce? Does it force loading the entirety of RxJS in the browser? I don't know how to check/measure this at this point. If RxJS was extracted from redux-observable, it might alleviate concerns of people wanting to start in a "light" way and only gradually bundle what they actually need. Would you have some size numbers to share, e.g in comparison to redux-saga (even if it's a bit of an apples-oranges thing)?

**Comment by BerkeleyTrue** [comment]
@sedubois RO only imports a subset of RxJS

**Comment by jayphelps** [comment]
@sebald yeah as Berkley says, redux-observable only imports what it needs to provide its functionality, so not all of RxJS.

I haven't compared bundle size to redux-saga because it would be really really deceptive because we don't import operators you'll likely import yourself. I would strongly recommend not picking one over the other on file size, or using a different stream library for that reason.

**Comment by bali182** [comment]
I looked through the code, and found that these are the things directly used from `rxjs`

- `Observable` - Would need some clever adaptation
- `Subject` - Same as the base `Observable`
- `of` - could be probably avoided if `ActionsObservable` is not used directly, but we need some kind of "constructor" for `Observables`
- `from` - same as `of`

And the operators:

- `filter` - wouldn't be necessary if `ActionsObservable` is not used directly
- `map` - seems necessary, but could be `re-implemented`
- `switchMap` - same as `map`

So basically we are left with some kind of `Observable`/`Subject` abstraction, that has similar `onNext`/`onCompleted`/`onError` internals as the rx implementation + has a convenient way of creation. Any idea what abstraction would be fitting for this?

**Comment by jayphelps** [comment]
@bali182 sorry, I didn't include an example of what the adapter would look like cause it's pretty simple--same as [existing adapters](https://github.com/redux-observable/redux-observable-adapter-rxjs-v4/blob/master/src/index.js#L33-L36)

A hypothetical rxjs v5 adapter would be something like

```js
class ActionsObservable { /* stuff */ }
const combineEpics = (...epics) => { /* stuff */ };

export default {
  input: input$ => ActionsObservable.from(input$),
  output: output$ => output$ // no change needed cause rxjs has Symbol.observable
};
```

It would be published to something like `@redux-observable/rxjs`, and used like this by consumers:

```js
import rxjsAdapter, { combineEpics } from '@redux-observable/rxjs';

const rootEpic = combineEpics(epic1, epic2);
const epicMiddleware = createEpicMiddleware(rootEpic, { adapter: rxjsAdapter });
```

**Comment by staltz** [comment]
By the way, we finished the release candidate for Cycle Unified https://github.com/cyclejs/cyclejs/releases/tag/unified-tag which removes the "adapters" we had before and depends on ES Observable interop.
